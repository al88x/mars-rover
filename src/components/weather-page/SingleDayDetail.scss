@import '../../style/constants.scss';
@import './Weather-constants.scss';

$summary-width: 60px;
$summary-margin: 2px;

$summary-total-width: ($summary-width * 7);

.SingleDayDetail {
    @include flex-row;
    @include box-styling;

    align-items: center;
    height: $detail-height * 3;
              
    @include svg-icon;    

    .weekdayBlock {  
        @include detail-styles;    
        margin-left: 10px;                
    }

    .tempBlock {
        @include detail-styles;       
        text-align: center;
    }  
    .windBlock {
        @include detail-styles;       
        text-align: center;
    } 
    
    .textBlock {
        @include detail-styles;      
        text-align: center;
        margin-right: 10px;
    } 
}
