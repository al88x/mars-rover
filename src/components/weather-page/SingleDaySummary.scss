@import '../../style/constants.scss';
@import './Weather-constants.scss';

.SingleDaySummary {
    @include box-styling;
    @include flex-column;   
    @include font-styles;     
   
    margin-top: 8px;

    width: $summary-width;
    height: $detail-height * 3;    
   
    text-align: center;
    color: $color-foreground;

    .weekday, .averageTemp {   
        @include summary-styles;                
    }

    .icon {
        @include summary-styles; 
        @include svg-icon;
    }

    .selection-highlight { 
        min-height: 8px;   
    }

    &:hover {
        @include box-gradient-hover;          
        .selection-highlight {        
            background-color: $color-red;
        }
    }    
}