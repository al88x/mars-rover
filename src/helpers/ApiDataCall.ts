import moment from "moment";

export type SingleDayData = {
    id:number;
    marsSolDate?: string;
    earthSolDate?: string
    averageTemp?: number;
    minTemp?: number;
    maxTemp?: number;
    averageWindSpeed?: number;
    windDirection?: string;
    averagePressure?: number;
    weekdayName: string;
}

export async function fetchWeatherData(): Promise<SingleDayData[]> {
    const response = await fetch('https://api.nasa.gov/insight_weather/?api_key=dxcDht667Acmm4X2g4KpXub7NK5hU2S8GqL8Y067&feedtype=json&ver=1.0');
    const json = await response.json();
    return parseData(json);
} 

function parseData(apiJson: any): SingleDayData[] {
    const dayData: SingleDayData[] = [];
    const solKeys : string[] = apiJson['sol_keys'];
    let identity :number = 0;
    solKeys.forEach(key => {
        const apiData = apiJson[key];
        const data: SingleDayData = {  
            id:identity,
            marsSolDate: key,
            earthSolDate: moment(apiData.First_UTC).format("MMM Do YY"),            
            averageTemp: apiData.AT.av,
            minTemp: apiData.AT.mn,
            maxTemp: apiData.AT.mx,
            averageWindSpeed: apiData.HWS.av,
            windDirection: apiData.WD.most_common.compass_point,
            averagePressure: apiData.PRE.av,
            weekdayName: moment(apiData.First_UTC).format('dddd').substring(0,3)
        }
        dayData.push(data)
        identity++;
    })
    return dayData;
}