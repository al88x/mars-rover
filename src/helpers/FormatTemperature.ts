export default function formatTemperature (temperature?: number) : string{
    if(!temperature){
        return "N/A";
    }
    return Math.round(temperature) + ' Â°C';
}