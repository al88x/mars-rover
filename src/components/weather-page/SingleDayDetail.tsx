import React from 'react';
import {SingleDayData} from './SingleDayData';
import './SingleDayDetail.scss';

import formatTemperature from '../../helpers/FormatTemperature';

import ThermometerIcon from './resources/svg_icon_thermometer';
import CloudIcon from './resources/svg_icon_cloud';
import SunIcon from './resources/svg_icon_sun';
import WindIcon from './resources/svg_icon_wind';

export default function SingleDayDetail(props : SingleDayData) {  

    const min = formatTemperature(props.minTemp);
    const max = formatTemperature(props.maxTemp);
    
    return (
        <div className="SingleDayDetail">            

            <div className="weekdayBlock">
                <p>DATE</p>
                <p>{props.weekdayName} {props.earthSolDate}</p>
                <p>MARS SOL {props.marsSolDate}</p>  
            </div> 

            <div className="tempBlock">                
                <p>High</p> 
                <p>{max}</p>
                <p>Low</p> 
                <p>{min}</p>
            </div>

            <div className="windBlock">
                <WindIcon/>  
                <p>{props.averageWindSpeed} mph</p>
                <p>{props.windDirection}</p>
            </div>

            <p className="textBlock">
                "Sunny with a chance of meatballs."
            </p>         
                      
        </div>
    )
}