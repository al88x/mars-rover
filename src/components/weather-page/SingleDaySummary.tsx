import React from 'react';
import {SingleDayData} from './SingleDayData';
import './SingleDaySummary.scss';

import ThermometerIcon from './resources/svg_icon_thermometer';
import CloudIcon from './resources/svg_icon_cloud';
import SunIcon from './resources/svg_icon_sun';
import WindIcon from './resources/svg_icon_wind';

// TODO - Add Image Array & random picker function for weather card icon
// TODO - Add function to convert "2019-11-14T16:08:54Z" into weekday i.e. Monday

export function SingleDaySummary(props : SingleDayData) {

    return (
        <div className="SingleDaySummary">

            <div className="DayIconTempCluster">

                <div className="DayNameContainer">
                    <h2>{props.weekdayName}</h2>
                </div>

                <div className="IconContainer">
                    <WindIcon/> 
                </div>
                
                <div className="TempContainer">
                    <div>{props.minTemp}</div>
                    <div>{props.maxTemp}</div>
                </div>                

            </div>

            <div className="TextSummary">
                <p>
                    Lorem Ipsum is simply dummy text of the printing and typesetting industry. 
                    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
                </p>
            </div>
        </div>
    )
}