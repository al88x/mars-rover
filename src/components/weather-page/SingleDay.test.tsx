import React from 'react';
import Renderer from 'react-test-renderer/shallow';
import SingleDayDetail from './SingleDayDetail';
import SingleDaySummary from './SingleDaySummary';
import { SingleDayData } from '../../helpers/ApiDataCall';

describe("Single Day Summary Component", () => {
    it("Should render without errors", () => {
        const testData: SingleDayData = {  
            id:1,
            marsSolDate: '344',
            earthSolDate: '14th',
            averageTemp: -67.108,
            minTemp: -99.926,
            maxTemp: -23.287,
            averageWindSpeed: 5.3,
            windDirection: 'SSE',
            averagePressure: 681.448,
            weekdayName: 'Monday' // TODO calc via function
        }

        const renderer = new Renderer();
        const component = renderer.render(
            <SingleDaySummary data = {testData}/>
        );
        
        expect(component).toMatchSnapshot();
    });
});

describe("Single Day Detail Component", () => {
    it("Should render without errors", () => {
        const testData: SingleDayData = {  
            id:1,
            marsSolDate: '344',
            earthSolDate: '14th',
            averageTemp: -67.108,
            minTemp: -99.926,
            maxTemp: -23.287,
            averageWindSpeed: 5.3,
            windDirection: 'SSE',
            averagePressure: 681.448,
            weekdayName: 'Monday' // TODO calc via function
        }

        const renderer = new Renderer();
        const component = renderer.render(
            <SingleDayDetail data = {testData}/>
        );
        
        expect(component).toMatchSnapshot();
    });
});