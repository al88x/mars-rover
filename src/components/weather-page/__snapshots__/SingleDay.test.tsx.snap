// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single Day Detail Component Should render without errors 1`] = `
<div
  className="SingleDayDetail"
>
  <div
    className="weekdayBlock"
  >
    <p>
      DATE
    </p>
    <p>
      Monday
       
      14th
    </p>
    <p>
      MARS SOL 
      344
    </p>
  </div>
  <div
    className="tempBlock"
  >
    <p>
      High
    </p>
    <p>
      -23 째C
    </p>
    <p>
      Low
    </p>
    <p>
      -100 째C
    </p>
  </div>
  <div
    className="windBlock"
  >
    <WindIcon />
    <p>
      5.3
       mph
    </p>
    <p>
      SSE
    </p>
  </div>
  <p
    className="textBlock"
  >
    "Sunny with a chance of meatballs."
  </p>
</div>
`;

exports[`Single Day Summary Component Should render without errors 1`] = `
<div
  className="SingleDaySummary"
>
  <div
    className="selection-highlight"
  />
  <div
    className="weekday"
  >
    Monday
  </div>
  <div
    className="icon"
  >
    <SunIcon />
  </div>
  <div
    className="averageTemp"
  >
    -67 째C째C
  </div>
</div>
`;
